File: struct/disjoint-set.cc
#include <vector>
using std::vector;

/**
 * 带路径压缩的并查集
 **/
class DisjointSet {
public:
    DisjointSet(int n){
        p = vector<int>(n + 1);
        rank = vector<int>(n + 1);
        for (int i = 0; i <= n; ++i) {
            p[i] = i;
            rank[i] = 0;
        }
    }
    bool inSame(int u, int v) {
        return findSet(u) == findSet(v);
    }
    void join(int u, int v) {
        if (findSet(u) != findSet(v)) {
            link(findSet(u), findSet(v));
        }
    }
    int findSet(int u) {
        if (u != p[u]) {
            p[u] = findSet(p[u]);
        }
        return p[u];
    }
private:
    vector<int> p;
    vector<int> rank;
    void link(int u, int v) {
        if (rank[u] > rank[v]) {
            p[v] = u;
        } else {
            p[u] = v;
            if (rank[u] == rank[v]) {
                rank[u] = rank[v] + 1;
            }
        }
    }
};

File: struct/disjoint-set.go
type DisjointSet struct {
	P    []int
	Rank []int
}

func InitDisjointSet(n int) *DisjointSet {
	d := &DisjointSet{}
	d.P = make([]int, n)
	d.Rank = make([]int, n)
	for i := 0; i < n; i++ {
		d.P[i] = i
	}
	return d
}

func (d *DisjointSet) InSame(u, v int) bool {
	return d.FindSet(u) == d.FindSet(v)
}

func (d *DisjointSet) Join(u, v int) {
	if d.FindSet(u) != d.FindSet(v) {
		d.Link(d.FindSet(u), d.FindSet(v))
	}
}

func (d *DisjointSet) FindSet(u int) int {
	if u != d.P[u] {
		d.P[u] = d.FindSet(d.P[u])
	}
	return d.P[u]
}

func (d *DisjointSet) Link(u, v int) {
	if d.Rank[u] > d.Rank[v] {
		d.P[v] = u
	} else {
		d.P[u] = v
		if d.Rank[u] == d.Rank[v] {
			d.Rank[u] = d.Rank[v] + 1
		}
	}
}
